### PR TITLE
feat: Support OAuth2 TLSConfig and ProxyConfig settings

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Fix: #6717 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Support OAuth2 TLSConfig and ProxyConfig settings 
```

### CODE REVIEW

**Comment by dongjiang1989** [comment]
This PR depend on #6726 . Wait for #6726 to be merged first.

File: pkg/apis/monitoring/v1/types.go [hunk 1]
```diff
@@ -565,6 +565,14 @@ type OAuth2 struct {
 	//
 	// +optional
 	EndpointParams map[string]string `json:"endpointParams,omitempty"`
+
+	// TLS configuration to use when scraping the target.
+	//
+	// +optional
+	TLSConfig *TLSConfig `json:"tlsConfig,omitempty"`
```
Comment (dongjiang1989) [inline pkg/apis/monitoring/v1/types.go:572]: use `TLSConfig` or `SafeTLSConfig` ? 🤔️

Comment (slashpai) [reply pkg/apis/monitoring/v1/types.go:572]: SafeTLSConfig

Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/types.go:572]: Got it.

**Comment by dongjiang1989** [comment]
@simonpasquier @slashpai PTAL. thanks

**Comment by dongjiang1989** [comment]
`rate limited even after retry; status code 429: Too Many Requests`  
Please rebuild CI pipeline.  thanks

File: pkg/apis/monitoring/v1/types.go [hunk 2]
```diff
@@ -569,6 +569,17 @@ type OAuth2 struct {
 	//
 	// +optional
 	EndpointParams map[string]string `json:"endpointParams,omitempty"`
+
+	// TLS configuration to use when scraping the target.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:573]:
```suggestion
	// TLS configuration to use when connecting to the OAuth2 server.
```

File: pkg/apis/monitoring/v1/types.go [hunk 3]
```diff
@@ -569,6 +569,17 @@ type OAuth2 struct {
...
+	// It requires Prometheus >= v2.43.0.
+	//
+	// +optional
+	TLSConfig *SafeTLSConfig `json:"tlsConfig,omitempty"`
+
+	// It requires Prometheus >= v2.43.0.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:579]:
```suggestion
	// Proxy configuration to use when connecting to the OAuth2 server.
	//
	// It requires Prometheus >= v2.43.0.
```

**Review by simonpasquier — COMMENTED** [review]
Thanks for carrying on! You need to extend `StoreBuilder.AddOAuth2()` to resolve the references to proxy connect headers + TLS cert, key and CA.
Alertmanager would need to be covered too (or it's deferred to a follow-up PR but in this case, it should be noted in the documentation)

File: pkg/apis/monitoring/v1/types.go (hunk 2 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/types.go]: Thanks. Done.

File: pkg/apis/monitoring/v1/types.go (hunk 3 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/types.go:580]: Thanks. Done.

**Comment by dongjiang1989** [comment]
> Thanks for carrying on! You need to extend `StoreBuilder.AddOAuth2()` to resolve the references to proxy connect headers + TLS cert, key and CA. Alertmanager would need to be covered too (or it's deferred to a follow-up PR but in this case, it should be noted in the documentation)

Thanks. @simonpasquier  Add `StoreBuilder.AddOAuth2()` done.
and `Alertmanager` OAuth2 covered for a follow-up PR.

File: pkg/apis/monitoring/v1/types.go [hunk 4]
```diff
@@ -569,6 +569,18 @@ type OAuth2 struct {
...
 	EndpointParams map[string]string `json:"endpointParams,omitempty"`
+
+	// TLS configuration to use when connecting to the OAuth2 server.
+	// It requires Prometheus >= v2.43.0.
+	//
+	// +optional
+	TLSConfig *SafeTLSConfig `json:"tlsConfig,omitempty"`
+
+	// Proxy configuration to use when connecting to the OAuth2 server.
+	// It requires Prometheus >= v2.43.0.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:580]:
```suggestion
	// It requires Prometheus >= v2.43.0.
	// It is not supported yet for Alertmanager.
```

**Comment by simonpasquier** [comment]
> Add `StoreBuilder.AddOAuth2()` done.

Hmm I don't see that change.

File: pkg/assets/store.go [hunk 5]
```diff
@@ -138,6 +138,16 @@ func (s *StoreBuilder) AddOAuth2(ctx context.Context, ns string, oauth2 *monitor
...
+	err = s.AddProxyConfig(ctx, ns, oauth2.ProxyConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get oauth2 proxyConfig: %w", err)
+	}
+
+	err = s.AddSafeTLSConfig(ctx, ns, oauth2.TLSConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get oauth2 tlsConfig: %w", err)
+	}
+
```
Comment (dongjiang1989) [inline pkg/assets/store.go:150]: @simonpasquier  AddOAuth

File: pkg/apis/monitoring/v1/types.go (hunk 4 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/types.go:580]: Thanks. Done.

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Comment by zwilling79** [comment]
I wanted to update the Prometheus Operator from v0.75.2 to v0.76.0 but the pod goes into a crash loop because of this error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1855ee2]
goroutine 301 [running]:
github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1.(*SafeTLSConfig).Validate(0x0)
	/home/runner/work/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1/types.go:713 +0x22
github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1.(*OAuth2).Validate(0x414518?)
	/home/runner/work/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1/types.go:610 +0x10c
github.com/prometheus-operator/prometheus-operator/pkg/assets.(*StoreBuilder).AddOAuth2(0xc0011c2600, {0x29c3c90, 0xc0007081e0}, {0xc0009234f6, 0x7}, 0xc000dd0240)
	/home/runner/work/prometheus-operator/prometheus-operator/pkg/assets/store.go:127 +0x73
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.(*ResourceSelector).SelectServiceMonitors(0xc000b3b220, {0x29c3c90, 0xc0007081e0}, 0xc000a0fc48)
	/home/runner/work/prometheus-operator/prometheus-operator/pkg/prometheus/resource_selector.go:170 +0x1467
github.com/prometheus-operator/prometheus-operator/pkg/prometheus/server.(*Operator).createOrUpdateConfigurationSecret(0xc0002e9908, {0x29c3c90, 0xc0007081e0}, 0xc00107a708, 0xc000dcbef0, {0xc000cd7cd0, 0x1, 0x1}, 0xc0011c2600)
	/home/runner/work/prometheus-operator/prometheus-operator/pkg/prometheus/server/operator.go:1138 +0x12d
github.com/prometheus-operator/prometheus-operator/pkg/prometheus/server.(*Operator).sync(0xc0002e9908, {0x29c3c90, 0xc0007081e0}, {0xc000aaccc0, 0x26})
	/home/runner/work/prometheus-operator/prometheus-operator/pkg/prometheus/server/operator.go:821 +0x6ac
github.com/prometheus-operator/prometheus-operator/pkg/prometheus/server.(*Operator).Sync(0xc0002e9908, {0x29c3c90?, 0xc0007081e0?}, {0xc000aaccc0, 0x26})
	/home/runner/work/prometheus-operator/prometheus-operator/pkg/prometheus/server/operator.go:767 +0x2c
github.com/prometheus-operator/prometheus-operator/pkg/operator.(*ResourceReconciler).processNextReconcileItem(0xc0002b2200, {0x29c3c90, 0xc0007081e0})
	/home/runner/work/prometheus-operator/prometheus-operator/pkg/operator/resource_reconciler.go:443 +0x231
github.com/prometheus-operator/prometheus-operator/pkg/operator.(*ResourceReconciler).Run.func1()
	/home/runner/work/prometheus-operator/prometheus-operator/pkg/operator/resource_reconciler.go:406 +0x3f
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/home/runner/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 259
	/home/runner/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:75 +0x96
```

I'm wondering whether it's because of this newly introduced feature? I'm using ServiceMonitor objects without  `oauth2.tlsConfig` settings in the cluster.

**Pull request merged**